test/ngRoute/routeSpec.js

```diff
@@ -2082,4 +2082,152 @@ describe('$route', function() {
+        $location.path('/path');
+        $rootScope.$digest();
+
+        var callback = jasmine.createSpy('callback');
+        $$testability.whenStable(callback);
+        expect(callback).not.toHaveBeenCalled();
+
+        deferred.reject();
+        $rootScope.$digest();
+        expect(callback).toHaveBeenCalled();
+      });
+    });
+
+    it('should wait for resolveRedirectTo promises before calling callbacks', function() {
+      var deferred;
+
+      module(function($provide, $routeProvider) {
+        $routeProvider.when('/path', { template: '', resolveRedirectTo: function($q) {
+            deferred = $q.defer();
+            return deferred.promise;
+          }
+        });
+      });
+
+      inject(function($location, $route, $rootScope, $httpBackend, $$testability) {
+        $location.path('/path');
+        $rootScope.$digest();
+
+        var callback = jasmine.createSpy('callback');
+        $$testability.whenStable(callback);
+        expect(callback).not.toHaveBeenCalled();
+
+        deferred.resolve();
+        $rootScope.$digest();
+        expect(callback).toHaveBeenCalled();
+      });
+    });
+
+    it('should call callback after resolveRedirectTo promises are rejected', function() {
+      var deferred;
+
+      module(function($provide, $routeProvider) {
+        $routeProvider.when('/path', { template: '', resolveRedirectTo: function($q) {
+            deferred = $q.defer();
+            return deferred.promise;
+          }
+        });
+      });
+
+      inject(function($location, $route, $rootScope, $httpBackend, $$testability) {
+        $location.path('/path');
+        $rootScope.$digest();
+
+        var callback = jasmine.createSpy('callback');
+        $$testability.whenStable(callback);
+        expect(callback).not.toHaveBeenCalled();
+
+        deferred.reject();
+        $rootScope.$digest();
+        expect(callback).toHaveBeenCalled();
+      });
+    });
+
+    it('should wait for all route promises before calling callbacks', function() {
+      var redirectDeferred;
+      var localsDeferred;
+
+      module(function($provide, $routeProvider) {
+        $routeProvider.
+          when('/foo', { template: '', resolveRedirectTo: function($q) {
+            redirectDeferred = $q.defer();
+            return redirectDeferred.promise;
+          } }).
+          when('/bar', { template: '', resolve: {
+            a: function($q) {
+              localsDeferred = $q.defer();
+              return localsDeferred.promise;
+            }
+          } });
+      });
+
+      inject(function($browser, $location, $route, $rootScope, $httpBackend, $$testability) {
+        $location.path('/foo');
+        $rootScope.$digest();
+
+        var callback = jasmine.createSpy('callback');
+        $$testability.whenStable(callback);
+        expect(callback).not.toHaveBeenCalled();
+
+        // ngRoute code is wrapped in a $browser.defer call (via $rootScope.evalAsync), which in
```

Comment:
Answering my own question: It comes from `$location` (see [here][1]).
So, you were spot on 👍 (which is a good thing, because it confirms that the feature works out-of-the-box outside of unit tests).

[1]: https://github.com/angular/angular.js/blob/752f411c8b3495620226e45815807ff798496801/src/ng/location.js#L991